simple/msg_data_integrity.c

```diff
@@ -0,0 +1,717 @@
+static int max_credits = 128;
+static int credits = 128;
+static char test_name[10] = "custom";
+static struct timespec start, end;
+static void *recv_buf;
+static void *send_buf;
+static size_t buffer_size;
+static size_t next_size = 16;
+static size_t transfer_size;
+
+static struct fi_info *hints;
+
+static struct fid_fabric *fab;
+static struct fid_pep *pep;
+static struct fid_domain *dom;
+static struct fid_ep *ep;
+static struct fid_eq *cmeq;
+static struct fid_cq *rcq, *scq;
+static struct fid_mr *recv_mr;
+static struct fid_mr *send_mr;
+static int common_seed = 7;
+static const char msg_char[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static int msg_char_length = (sizeof(msg_char)/sizeof(*msg_char)) - 1;
+static int allowed_error_count = 0;
+static int detailed_output = 0;
+static int detailed_error = 0;
+static int error_count = 0;
+static int first_error_byte = -1;
+static int print_error = 0;
+
+static void print_byte_array(char *buffer, char *array, int size)
+{
+	int j = 0;
+	printf("%s = ", buffer);
+	for(j=0; j<size; j++) {
+		printf("%c", array[j]);
+        }
+        printf("\n");
+}
+
+static int send_xfer(int size, int iter)
+{
+	struct fi_cq_entry comp;
+	int ret;
+	int j;
+	int msg_index;
+	char *msg_buf;
+
+	if (iter >= 0) {
+		msg_index = (iter*common_seed) % msg_char_length;
+	} else {
+		msg_index = 0;
+	}
+	while (!credits) {
+		ret = fi_cq_read(scq, &comp, 1);
+		if (ret > 0) {
+			goto post;
+		} else if (ret < 0 && ret != -FI_EAGAIN) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(scq, "scq");
+			} else {
+				FT_PRINTERR("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	}
+
+	credits--;
+post:
+	msg_buf = (char *)send_buf;
+
+	for (j = 0; j < size; j++) {
+		msg_buf[j] = msg_char[msg_index++];
+		if (msg_index >= msg_char_length) {
+			msg_index = 0;
+		}
+	}
+
+	ret = fi_send(ep, send_buf, (size_t) size, fi_mr_desc(send_mr), 0, NULL);
+	if (ret)
+		FT_PRINTERR("fi_send", ret);
+
+
+	if (detailed_output) {
+		print_byte_array("Send Buffer", msg_buf, size);
+	}
+
+	return ret;
+}
+
+static int recv_xfer(int size, int iter)
+{
+	struct fi_cq_entry comp;
+	char *recv_data;
+	char *expected_data;
+	int msg_index;
+	int ret;
+	int j = 0;
+
+	if (iter >= 0) {
+                msg_index = (iter*common_seed) % msg_char_length;
+        } else {
+                msg_index = 0;
+        }
+
+	expected_data = (char *) malloc(size);
+        if (!expected_data) {
+                perror("malloc");
+                return -1;
+        }
+
+        for (j = 0; j < size; j++) {
+                expected_data[j] = msg_char[msg_index++];
+                if (msg_index >= msg_char_length) {
+                        msg_index = 0;
+                }
+        }
+
+	do {
+		ret = fi_cq_read(rcq, &comp, 1);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(rcq, "rcq");
+			} else {
+				FT_PRINTERR("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	} while (ret == -FI_EAGAIN);
```

Comment:
I know this was probably cribbed from `msg_pingpong.c`, but you might consider using `fi_cq_sread` in order to ditch the loop.